libc/stdlib_strto.c

```diff
@@ -0,0 +1,671 @@
+{
+/* Disabled because of #704 issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/704 */
+#ifdef __phoenix__
+	TEST_IGNORE_MESSAGE("#704 issue");
+#endif
+	const char *str[] = { "Inf", "-Inf", "NaN", "-NaN" };
+	const double exp_double[] = { INFINITY, -INFINITY, NAN, -NAN };
+	const float exp_float[] = { INFINITY, -INFINITY, NAN, -NAN };
+	const long double exp_ldoubl[] = { INFINITY, -INFINITY, NAN, -NAN };
+
+	for (int i = 0; i < (sizeof(str) / sizeof(str[0])); i++) {
+		errno = 0;
+		TEST_ASSERT_EQUAL_DOUBLE(exp_double[i], strtod(str[i], NULL));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_FLOAT(exp_float[i], strtof(str[i], NULL));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+
+		errno = 0;
+		TEST_ASSERT_EQUAL_DOUBLE(exp_ldoubl[i], strtold(str[i], NULL));
+		TEST_ASSERT_EQUAL_INT(0, errno);
+	}
+}
+
+
+TEST(stdlib_strto, strto_endptr)
+{
+	const char str_float[] = "1.23 4.56";
+	const char str_int[] = "12345 67890";
+	char *end = NULL;
+
+	errno = 0;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_DOUBLE(1.23, strtod(str_float, &end));
+	TEST_ASSERT_EQUAL_DOUBLE(4.56, strtod(end, NULL));
+	TEST_ASSERT_EQUAL_PTR(str_float + 4, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_FLOAT(1.23, strtof(str_float, &end));
+	TEST_ASSERT_EQUAL_FLOAT(4.56, strtof(end, NULL));
+	TEST_ASSERT_EQUAL_PTR(str_float + 4, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_DOUBLE(1.23, strtold(str_float, &end));
+	TEST_ASSERT_EQUAL_DOUBLE(4.56, strtold(end, NULL));
+	TEST_ASSERT_EQUAL_PTR(str_float + 4, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_INT(12345, strtol(str_int, &end, 10));
+	TEST_ASSERT_EQUAL_INT(67890, strtol(end, NULL, 10));
+	TEST_ASSERT_EQUAL_PTR(str_int + 5, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_INT(12345, strtoll(str_int, &end, 10));
+	TEST_ASSERT_EQUAL_INT(67890, strtoll(end, NULL, 10));
+	TEST_ASSERT_EQUAL_PTR(str_int + 5, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_INT(12345, strtoul(str_int, &end, 10));
+	TEST_ASSERT_EQUAL_INT(67890, strtoul(end, NULL, 10));
+	TEST_ASSERT_EQUAL_PTR(str_int + 5, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	errno = 0;
+	end = NULL;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_INT(12345, strtoull(str_int, &end, 10));
+	TEST_ASSERT_EQUAL_INT(67890, strtoull(end, NULL, 10));
+	TEST_ASSERT_EQUAL_PTR(str_int + 5, end);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+}
+
+
+TEST(stdlib_strto, strto_empty)
+{
+	const char str[] = "";
+	char *end = NULL;
+
+	errno = 0;
+	TEST_ASSERT_NULL(end);
+	TEST_ASSERT_EQUAL_DOUBLE(0, strtod(str, &end));
+	TEST_ASSERT_EQUAL_PTR(str, end);
+	TEST_ASSERT_TRUE(errno == EINVAL || errno == 0);
```

Comment:
When does it return 0? I suppose that EINVAL should be setas errno.

Reply:
According to POSIX: When "No conversion could be performed", these functions may set errno to EINVAL.  So errno in these situations can be either 0 or EINVAL. 
![image](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/116297248/c7542783-29ed-47de-9a18-8d32623c84f0)

Reply:
This is checked in the following test: ` TEST(stdlib_strto, invalid_base)`